src/parser/jobs/blm/Leylines.js

```diff
@@ -0,0 +1,100 @@
+import React from 'react'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import Module from 'parser/core/Module'
+import {ActionLink} from 'components/ui/DbLink'
+import {i18nMark, Trans} from '@lingui/react'
+
+const LEYLINE_DURATION = 30000
+
+export default class Leylines extends Module {
+	static handle = 'leylines'
+	static i18n_id = i18nMark('blm.leylines.title')
+	static title = 'Ley Lines'
+
+	static dependencies = [
+		'checklist',
+	]
+	_CircleOfPowers = {
+		current: null,
+		history: [],
+	}
+	_LeyLineHistory = []
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('cast', {
+			by: 'player',
+			abilityId: ACTIONS.LEY_LINES.id,
+		}, this._onCastLeyLines)
+		this.addHook('applybuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onApplyCircleOfPower)
+		this.addHook('removebuff', {
+			by: 'player',
+			abilityId: STATUSES.CIRCLE_OF_POWER.id,
+		}, this._onRemoveCircleOfPower)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCastLeyLines(event) {
+		this.ll = this._LeyLineHistory
+		this.ll.push(event.timestamp)
+
+	}
+
+	_onApplyCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) { this.cop.history.push(this.cop.current) }
+
+		this.cop.current = {
+			start: event.timestamp,
+			stop: null,
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_onRemoveCircleOfPower(event) {
+		this.cop = this._CircleOfPowers
+		if (this.cop.current) {
+			this.cop.current.stop = event.timestamp
+		}
+		this._CircleOfPowers = this.cop
+	}
+
+	_percentFunction(numberOfLeyLines, SumOfCoPUpime) {
+		return (SumOfCoPUpime/(numberOfLeyLines*LEYLINE_DURATION))*100
+	}
+
+	_onComplete(event) {
+		this.cop = this._CircleOfPowers
+		if (!this.cop.history.length) { return }
+		if (this.cop.current) {
+			if (!this.cop.current.stop) {
+				this.cop.current.stop = event.timestamp
+			}
+			this.cop.history.push(this.cop.current)
+		}
+		//filter out the last possible LL usage because it would make things weird.
+		this.cop.history.filter(cops => cops.start < (event.timestamp - LEYLINE_DURATION))
+		this._LeyLineHistory.filter(timestamps => timestamps < (event.timestamp - LEYLINE_DURATION))
```

Comment:
I'm not sure what leaving them in does to your results, but filtering out LL events that run into the end of the fight have the potential to give you >100% uptime results. If, say, you had 100% uptime on all your mid-fight LL casts and cast your last one with ~20s left in the fight, stood in it for 15s, then moved out, that CoP event wouldn't get cut but the LL would, leaving you with more time in CoP than time with LL active. Not a super major issue since it's kind of an edge case, but worth considering.

Reply:
I also filter out all CoP events (given that the filter works) for exactly that reason

Reply:
Ah, but that's the thing - you're only filtering CoP events that hit the end of the fight. If one of them ends 5s before the fight ends, it won't get cut (which is the scenario I outlined above).